<a href="https://colab.research.google.com/github/saheli56/ml-handbook/blob/main/Decision%20Trees%20for%20Classification%20in%20Machine%20Learning%20-%20Practical%20-%201ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
dataset =pd.read_csv("/content/Social_Network_Ads.csv")

In [ ]:
dataset

In [ ]:
x= dataset.iloc[:,[2,3]].values
y = dataset.iloc[:,4].values

In [ ]:
x,y

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train,y_test = train_test_split(x,y,test_size=0.25,random_state=0)

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
sc = StandardScaler()

In [ ]:
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

In [ ]:
#gini - Impurity--range from 0-1--> 0 means pure (only one class)--1 means many classes and records (data points) Belong to multiple classes
# Purchased - output
# Parson - buy-1 or nod by =0
# Purely - one class
# Class - 0 and 1

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
dtclr = DecisionTreeClassifier(criterion='gini',splitter="random",random_state=0)

In [ ]:
dclr.fit(x_train,y_train)

In [ ]:
y_pred = dclr.predict(x_test)

In [ ]:
print(y_pred)

In [ ]:
print (y_test)

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix

In [ ]:
print("Accuracy score:",accuracy_score(y_test,y_pred))

In [ ]:
cf = confusion_matrix(y_test,y_pred)

In [33]:
from sklearn.metrics import confusion_matrix
cf = confusion_matrix(y_test,y_pred)
print(cf)

[[0 0 0 ... 0 0 0]
 [1 0 0 ... 0 0 0]
 [0 3 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 1]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [ ]:
!wget https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv -O /content/Social_Network_Ads.csv

--2025-09-08 15:19:18--  https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘/content/Social_Network_Ads.csv’

/content/Social_Net 100%[===================>]   1.36M  --.-KB/s    in 0.07s   

2025-09-08 15:19:18 (20.7 MB/s) - ‘/content/Social_Network_Ads.csv’ saved [1423529/1423529]



In [ ]:
# Train

In [ ]:
x= x_train[:,0]
y= x_train[:,1]
x,y

(array([-0.76426232, -0.84363104, -0.76426232, ...,  0.18816232,
         0.42626848, -1.0817372 ]),
 array([ 1.06809083, -0.48039972, -0.95569712, ..., -0.55359277,
        -0.2361179 , -0.67253149]))

# Task
Run the code in the notebook correctly to train and evaluate a Decision Tree Classifier on the provided dataset.

## Load and prepare data

### Subtask:
Load the dataset and extract the features (`x`) and target variable (`y`). Handle any missing values in `y` and ensure `x` remains a 2D array with consistent samples.


**Reasoning**:
The first step is to load the dataset into a pandas DataFrame as instructed.



In [26]:
dataset = pd.read_csv("/content/Social_Network_Ads.csv")

**Reasoning**:
Now that the dataset is loaded, I need to extract the features and target variable and handle any potential missing values in the target variable.



In [27]:
x = dataset.iloc[:, [2, 3]].values
y = dataset.iloc[:, 4].values

# Check for and handle missing values in y
if pd.isna(y).any():
    # A simple strategy is to drop rows with missing y values.
    # This assumes that the corresponding x values are also removed to maintain consistency.
    # A more sophisticated approach might involve imputation depending on the data and task.
    initial_len = len(y)
    non_nan_indices = ~pd.isna(y)
    x = x[non_nan_indices]
    y = y[non_nan_indices]
    if len(y) < initial_len:
        print(f"Removed {initial_len - len(y)} rows with missing target values.")

# Ensure x is a 2D array and check for consistent sample numbers
if x.ndim == 1:
    x = x.reshape(-1, 1)

if x.shape[0] != y.shape[0]:
    raise ValueError("Number of samples in features (x) and target (y) do not match.")

print("Features (x) shape:", x.shape)
print("Target (y) shape:", y.shape)

Removed 207 rows with missing target values.
Features (x) shape: (20433, 2)
Target (y) shape: (20433,)


## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Import the necessary function for splitting the data and then split the data into training and testing sets.



In [28]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=0)

## Feature scaling

### Subtask:
Scale the features using StandardScaler.


**Reasoning**:
Import StandardScaler, instantiate it, fit it to the training data, and transform both training and testing data.



In [29]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

## Train the model

### Subtask:
Initialize and train the Decision Tree Classifier.


**Reasoning**:
Import the DecisionTreeClassifier and instantiate it, then train the model using the scaled training data.



In [30]:
from sklearn.tree import DecisionTreeClassifier
dtclr = DecisionTreeClassifier(criterion='gini', splitter='random', random_state=0)
dtclr.fit(x_train, y_train)

DecisionTreeClassifier(random_state=0, splitter='random')

## Make predictions

### Subtask:
Use the trained model to make predictions on the test set.


**Reasoning**:
The previous cell failed because the variable `y_pred` was not defined. This subtask requires using the trained model `dtclr` to make predictions on the test set `x_test` and store them in `y_pred`.



In [31]:
y_pred = dtclr.predict(x_test)

## Evaluate the model

### Subtask:
Calculate and print the confusion matrix and accuracy score.


**Reasoning**:
Calculate and print the confusion matrix and accuracy score using the true and predicted labels.



In [32]:
from sklearn.metrics import confusion_matrix, accuracy_score
cf = confusion_matrix(y_test, y_pred)
print(cf)
print("Accuracy score:", accuracy_score(y_test, y_pred))

[[0 0 0 ... 0 0 0]
 [1 0 0 ... 0 0 0]
 [0 3 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 1]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
Accuracy score: 0.004306126443531024


## Summary:

### Data Analysis Key Findings

*   Initially, the dataset contained 207 rows with missing values in the target variable (`y`), which were successfully removed, resulting in a dataset of 20433 samples for both features and the target.
*   The dataset was split into training and testing sets, with the test set comprising 25% of the data.
*   Features were scaled using `StandardScaler` before training the model.
*   A Decision Tree Classifier was trained with `criterion='gini'` and `splitter='random'`.
*   The trained model achieved a very low accuracy score of 0.0043 on the test set.
*   The confusion matrix showed the distribution of true and predicted labels, further indicating the model's poor performance.

### Insights or Next Steps

*   The extremely low accuracy suggests that the chosen model and parameters are not suitable for this dataset, or there might be issues with the data or feature selection.
*   Investigate alternative models, hyperparameter tuning, or consider feature engineering and selection to improve performance.
